tensorflow/go/attrs.go

```diff
@@ -0,0 +1,258 @@
+				vals[i] = val == 1
+			}
+			return vals, nil
+
+		case TypeAttr:
+			list := make([]C.TF_DataType, a.ListSize)
+			C.TF_OperationGetAttrTypeList(a.Op.c, cname, &list[0], C.int(a.ListSize), status.c)
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			vals := make([]DataType, a.ListSize)
+			for i, val := range list {
+				vals[i] = DataType(val)
+			}
+			return vals, nil
+
+		case TensorAttr:
+			list := make([]*C.TF_Tensor, a.ListSize)
+			C.TF_OperationGetAttrTensorList(a.Op.c, cname, &list[0], C.int(a.ListSize), status.c)
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			vals := make([]*Tensor, a.ListSize)
+			for i, t := range list {
+				vals[i] = newTensorFromC(t)
+			}
+			return vals, nil
+
+		case ShapeAttr:
+			dims := make([]*C.int64_t, a.ListSize)
+			numDims := make([]C.int, a.ListSize)
+			storage := make([]C.int64_t, a.TotalSize)
+			C.TF_OperationGetAttrShapeList(a.Op.c, cname, &dims[0], &numDims[0], C.int(a.ListSize), &storage[0], C.int(a.TotalSize), status.c)
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			list := make([]Shape, a.ListSize)
+			for i, dim := range dims {
+				numDim := numDims[i]
+				// A []C.int64_t slice backed by C memory.
+				// See: https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
+				slice := (*[1 << 30]C.int64_t)(unsafe.Pointer(dim))[:numDim:numDim]
+				s := Shape{
+					dims: make([]int64, numDim),
+				}
+				for i, dim := range slice {
+					s.dims[i] = int64(dim)
+				}
+				list[i] = s
+			}
+			return list, nil
+
+		default:
+			return nil, errors.Errorf("list type %v not supported yet", a.Type)
+		}
+	}
+
+	switch a.Type {
+	case StringAttr:
+		v := make([]C.char, a.TotalSize)
+		C.TF_OperationGetAttrString(a.Op.c, cname, unsafe.Pointer(&v[0]), C.size_t(a.TotalSize), status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return C.GoStringN(&v[0], C.int(a.TotalSize)), nil
+
+	case IntAttr:
+		var v C.int64_t
+		C.TF_OperationGetAttrInt(a.Op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return int64(v), nil
+
+	case FloatAttr:
+		var v C.float
+		C.TF_OperationGetAttrFloat(a.Op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return float32(v), nil
+
+	case BoolAttr:
+		var v C.uchar
+		C.TF_OperationGetAttrBool(a.Op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return v == 1, nil
+
+	case TypeAttr:
+		var v C.TF_DataType
+		C.TF_OperationGetAttrType(a.Op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return DataType(v), nil
+
+	case TensorAttr:
+		var v *C.TF_Tensor
+		C.TF_OperationGetAttrTensor(a.Op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return newTensorFromC(v), nil
+
+	case ShapeAttr:
+		numDims := a.TotalSize
+		dims := make([]C.int64_t, numDims)
```

Comment:
Shouldn't we be returning a `Shape` object instead of an `int64` slice here? Since the attribute can represent only partially known shapes, and the `Shape` type is designed for that?

Reply:
We do return a shape. This dims array is what the C fills into and then we wrap it with a Shape.